Sources/TunnelKitWireGuardAppExtension/WireGuardTunnelProvider.swift

```diff
@@ -41,41 +47,41 @@ open class WireGuardTunnelProvider: NEPacketTunnelProvider {
         }
 
         configureLogging()
-
         // END: TunnelKit
 
         // Start the tunnel
-        adapter.start(tunnelConfiguration: tunnelConfiguration) { adapterError in
+        adapter.start(tunnelConfiguration: tunnelConfiguration) { [weak self] adapterError in
```

Comment:
This is fine, but if you then `guard let self` as the first line in the completion handler, you avoid all the subsequent modifications to use `self?` instead of `self`.

```suggestion
        adapter.start(tunnelConfiguration: tunnelConfiguration) { [weak self] adapterError in
            guard let self else {
                completionHandler(nil)
                return
            }
```

Reply:
Commonly I use optional chaning, if I am wrong, please let me know
Anyway I will fix this code snippet according to your preference

Reply:
Nothing "wrong", but I'd rather fail fast if a precondition is unmet.